File: solutions/python/say/1/say.py
def say(number):
    if number < 0 or number > 999_999_999_999:
        raise ValueError("input out of range")
    
    if number == 0:
        return "zero"

    ones = ["", "one", "two", "three", "four", "five", "six", "seven", "eight", "nine", "ten",
            "eleven", "twelve", "thirteen", "fourteen", "fifteen", "sixteen", "seventeen", "eighteen", "nineteen"]
    tens = ["", "", "twenty", "thirty", "forty", "fifty", "sixty", "seventy", "eighty", "ninety"]

    def say_under_100(n):
        if n < 20:
            return ones[n]
        else:
            tens_part = tens[n // 10]
            ones_part = ones[n % 10]
            if ones_part:
                return f"{tens_part}-{ones_part}"
            return tens_part

    def say_under_1000(n):
        hundreds = n // 100
        remainder = n % 100
        parts = []
        if hundreds > 0:
            parts.append(f"{ones[hundreds]} hundred")
        if remainder > 0:
            parts.append(say_under_100(remainder))
        return " ".join(parts)

    temp = number
    groups = []
    while temp > 0:
        groups.append(temp % 1000)
        temp //= 1000

    units = ["", "thousand", "million", "billion"]
    results = []
    for i, g in enumerate(groups):
        if g > 0:
            g_str = say_under_1000(g)
            unit = units[i]
            if unit:
                results.append(f"{g_str} {unit}")
            else:
                results.append(g_str)

    results.reverse()
    return " ".join(results)

File: solutions/ruby/say/1/say.rb
class Say
  def initialize(number)
    @number = number
  end

  def in_english
    raise ArgumentError if @number < 0 || @number >= 1_000_000_000_000
    return 'zero' if @number == 0

    result = ''
    chunks = chunk_number(@number)
    chunks.each_with_index.reverse_each do |chunk, i|
      next if chunk == 0
      result << ' ' unless result.empty?
      result << under_1000(chunk)
      result << " #{SCALE[i]}" if SCALE[i]
    end
    result
  end

  private

  SCALE = [nil, 'thousand', 'million', 'billion']

  def chunk_number(n)
    chunks = []
    while n > 0
      chunks << n % 1000
      n /= 1000
    end
    chunks
  end

  def under_1000(n)
    result = ''
    if n >= 100
      result << "#{ONES[n / 100]} hundred"
      n %= 100
      result << ' ' if n > 0
    end
    if n >= 20
      result << TENS[n / 10]
      n %= 10
      result << '-' if n > 0
    elsif n >= 10
      result << (n == 10 ? 'ten' : TEENS[n - 10])
      n = 0
    end
    if n > 0
      result << ONES[n]
    end
    result
  end

  ONES = [nil, 'one', 'two', 'three', 'four', 'five', 'six', 'seven', 'eight', 'nine']
  TEENS = [nil, 'eleven', 'twelve', 'thirteen', 'fourteen', 'fifteen', 'sixteen', 'seventeen', 'eighteen', 'nineteen']
  TENS = [nil, 'ten', 'twenty', 'thirty', 'forty', 'fifty', 'sixty', 'seventy', 'eighty', 'ninety']
end
